README.md
[![Gem Version](https://badge.fury.io/rb/dcc.png)](http://badge.fury.io/rb/dcc)
[![Build Status](https://secure.travis-ci.org/robertgauld/dcc.png?branch=main)](http://travis-ci.org/robertgauld/dcc)
[![Coveralls Status](https://coveralls.io/repos/robertgauld/dcc/badge.png?branch=main)](https://coveralls.io/r/robertgauld/dcc)
[![Code Climate](https://codeclimate.com/github/robertgauld/dcc.png?branch=main)](https://codeclimate.com/github/robertgauld/dcc)

## Ruby Versions
This gem supports the following versions of ruby, it may work on other versions but is not tested against them so don't rely on it.

* ruby:
  * 2.7.0 - 2.7.1


## DCC

Work with the NMRA's Digital Command Control standards.

Details of the gem's API can be found at <https://rubydoc.info/github/robertgauld/dcc/main>

## Installation

If you're using bundler then add it to your Gemfile and run the bundle command.

```ruby
gem 'dcc', '~> 0.1'
```

If you're not using bundler then install it from the command line.
```bash
gem install dcc -v '~> 0.1'
```

## Usage

### Setup

```ruby
require 'dcc'    # Unless you're using Bundler
```

### Examples

#### Work with long loco addresses
```ruby
# Get the values of CVs 17 and 18 for a given address
cv17, cv18 = Dcc.long_loco_address_bytes 1234
# it returns [196, 210] so cv17 is 196, and cv18 is 210

# Get an address given the values of CVs 17 and 18
Dcc.long_loco_address 196, 210
# returns 1234
```

#### Work with CV29

```ruby
# Make a value for writing to CV29
cv29 = Dcc.cv29 :speed_steps, :railcom, :long_address
# cv29 is 42

# Query a value read from CV29
Dcc.cv29? cv29, :reverse_direction   # returns false
Dcc.cv29? cv29, :speed_steps         # returns true
Dcc.cv29? cv29, :dc_operation        # returns false
Dcc.cv29? cv29, :railcom             # returns true
Dcc.cv29? cv29, :complex_speed_curve # returns false
Dcc.cv29? cv29, :long_address        # returns true
```

## Versioning

We follow the [Semantic Versioning](http://semver.org/) concept.


CHANGELOG.md
# DCC Gem Changelog

## Version 0.1.0

* Initial release.


spec/spec_helper.rb
# frozen_string_literal: true

require 'coveralls'
require 'rspec/its'
require 'simplecov'

SimpleCov.coverage_dir(File.join('tmp', 'coverage')) unless ENV.key?('TRAVIS')
SimpleCov.start do
  add_filter 'spec/'
end

Coveralls.wear! if ENV.key?('TRAVIS')

RSpec.configure do |config|
  # == Mock Framework
  #
  # If you prefer to use mocha, flexmock or RR, uncomment the appropriate line:
  #
  # config.mock_with :mocha
  # config.mock_with :flexmock
  # config.mock_with :rr
  config.mock_with :rspec  do |configuration|
    # Using the expect syntax is preferable to the should syntax in some cases.
    # The problem here is that the :should syntax that RSpec uses can fail in
    # the case of proxy objects, and objects that include the delegate module.
    # Essentially it requires that we define methods on every object in the
    # system. Not owning every object means that we cannot ensure this works in
    # a consistent manner. The expect syntax gets around this problem by not
    # relying on RSpec specific methods being defined on every object in the
    # system.
    # configuration.syntax = [:expect, :should]
    configuration.syntax = :expect
  end
end

require 'dcc'


spec/units/dcc_spec.rb
# frozen_string_literal: true

describe Dcc do
  describe '#long_loco_address' do
    it 'Calulates values' do
      expect(described_class.long_loco_address(196, 210)).to eq 1234
    end

    it 'Fails if CV17 is out of range' do
      expect { described_class.long_loco_address(-1, 128) }.to raise_error ArgumentError, 'cv17 is out of range (0-255)'
      expect { described_class.long_loco_address(256, 128) }.to raise_error ArgumentError, 'cv17 is out of range (0-255)'
    end

    it 'Fails if CV18 is out of range' do
      expect { described_class.long_loco_address(128, -1) }.to raise_error ArgumentError, 'cv18 is out of range (0-255)'
      expect { described_class.long_loco_address(128, 256) }.to raise_error ArgumentError, 'cv18 is out of range (0-255)'
    end
  end

  describe '#long_loco_address_bytes' do
    it 'Calculates address' do
      expect(described_class.long_loco_address_bytes(1234)).to eq [196, 210]
    end

    it 'Fails if address is out of range' do
      expect { described_class.long_loco_address_bytes(0) }
        .to raise_error ArgumentError, 'address is out of range (1-10239)'
      expect { described_class.long_loco_address_bytes(10_240) }
        .to raise_error ArgumentError, 'address is out of range (1-10239)'
    end
  end

  describe '#cv29' do
    describe 'Sets bit for' do
      it 'reverse_direction' do
        expect(described_class.cv29(:reverse_direction)).to eq 1
      end

      it 'speed_steps' do
        expect(described_class.cv29(:speed_steps)).to eq 2
      end

      it 'dc_operation' do
        expect(described_class.cv29(:dc_operation)).to eq 4
      end

      it 'railcom' do
        expect(described_class.cv29(:railcom)).to eq 8
      end

      it 'complex_speed_curve' do
        expect(described_class.cv29(:complex_speed_curve)).to eq 16
      end

      it 'long_address' do
        expect(described_class.cv29(:long_address)).to eq 32
      end
    end

    it 'Sets multiple bits' do
      expect(described_class.cv29(:speed_steps, :railcom, :long_address)).to eq 42
    end
  end

  describe '#cv29?' do
    describe 'Checks bit for' do
      it 'reverse_direction' do
        expect(described_class.cv29?(0b00000001, :reverse_direction)).to be true
        expect(described_class.cv29?(0b11111110, :reverse_direction)).to be false
      end

      it 'speed_steps' do
        expect(described_class.cv29?(0b00000010, :speed_steps)).to be true
        expect(described_class.cv29?(0b11111101, :speed_steps)).to be false
      end

      it 'dc_operation' do
        expect(described_class.cv29?(0b00000100, :dc_operation)).to be true
        expect(described_class.cv29?(0b11111011, :dc_operation)).to be false
      end

      it 'railcom' do
        expect(described_class.cv29?(0b00001000, :railcom)).to be true
        expect(described_class.cv29?(0b11110111, :railcom)).to be false
      end

      it 'complex_speed_curve' do
        expect(described_class.cv29?(0b00010000, :complex_speed_curve)).to be true
        expect(described_class.cv29?(0b11101111, :complex_speed_curve)).to be false
      end

      it 'long_address' do
        expect(described_class.cv29?(0b00100000, :long_address)).to be true
        expect(described_class.cv29?(0b11011111, :long_address)).to be false
      end
    end

    it 'Fails on invalid option' do
      expect { described_class.cv29?(255, :invalid) }.to raise_error ArgumentError, ':invalid is not a valid option'
    end
  end
end


lib/dcc.rb
# frozen_string_literal: true

require 'zeitwerk'

# Custom inflector for Zeitwerk.
# @api private
class MyInflector < Zeitwerk::Inflector
  # Convert file's base name to class name when
  # Zeitwerk's included inflector gets it wrong.
  # @param basename [String] the file's base name (no path or extension)
  # @param _abspath [String] the file's absolute path
  # @return [String] the class name
  def camelize(basename, _abspath)
    case basename
    when 'version' then 'VERSION'
    else
      super
    end
  end
end

loader = Zeitwerk::Loader.for_gem
loader.inflector = MyInflector.new
loader.setup

# DCC calculators (and namespeace for the gem).
module Dcc
  # Get a long loco address from the values of CVs 17 and 18.
  # @param cv17 [Integer] the value read from CV17 (0-255)
  # @param cv18 [Integer] the value read from CV18 (0-255)
  # @return [Integer] the long loco address resulting from the passed values
  def self.long_loco_address(cv17, cv18)
    fail ArgumentError, 'cv17 is out of range (0-255)' unless cv17.between? 0, 255
    fail ArgumentError, 'cv18 is out of range (0-255)' unless cv18.between? 0, 255

    ((cv17 - 192) * 256) + cv18
  end

  # Calculate the bytes to store in CVs 17 and 18 for a long loco address.
  # @param address [Integer] the long loco address (1-102390)
  # @return [Array<Integer, Integer>] the bytes for CVs 17 and 18 (in that order)
  def self.long_loco_address_bytes(address)
    fail ArgumentError, 'address is out of range (1-10239)' unless address.between? 1, 10_239

    [
      (address / 256).floor + 192,
      address % 256
    ]
  end

  # Calculate the value of CV 29 for the passed options.
  # @param options [Array<Symbol>] any combination of:
  #   :reverse_direction, :speed_steps, :dc_operation, :railcom,
  #   :complex_speed_curve or :long_address
  # @return [Integer] the byte for CV 29
  def self.cv29(*options)
    value = 0
    value += 1 if options.include? :reverse_direction
    value += 2 if options.include? :speed_steps
    value += 4 if options.include? :dc_operation
    value += 8 if options.include? :railcom
    value += 16 if options.include? :complex_speed_curve
    value += 32 if options.include? :long_address
    value
  end

  # Check if an option is set in a given value of CV 29.
  # @param value [Integer] the value read from CV 29
  # @param option [Symbol] either:
  #   :reverse_direction, :speed_steps, :dc_operation, :railcom,
  #   :complex_speed_curve or :long_address
  # @return [Boolean]
  def self.cv29?(value, option)
    case option
    when :reverse_direction   then (value & 1).positive?
    when :speed_steps         then (value & 2).positive?
    when :dc_operation        then (value & 4).positive?
    when :railcom             then (value & 8).positive?
    when :complex_speed_curve then (value & 16).positive?
    when :long_address        then (value & 32).positive?
    else
      fail ArgumentError, "#{option.inspect} is not a valid option"
    end
  end
end
